tensorflow/core/grappler/optimizers/meta_optimizer.cc

```diff
@@ -228,72 +253,104 @@ MetaOptimizer::MetaOptimizer(DeviceBase* cpu_device, const ConfigProto& cfg)
 }
 
 Status MetaOptimizer::InitializeOptimizers(
-    std::vector<std::unique_ptr<GraphOptimizer>>* optimizers) const {
+    std::vector<std::unique_ptr<GraphOptimizer>>* optimizers,
+    const std::set<string>& device_types) const {
   if (cfg_.disable_meta_optimizer()) {
     return Status::OK();
   }
-  if (!cfg_.disable_model_pruning()) {
+
+  ConfigList plugin_configs = PluginGraphOptimizerRegistry::GetPluginConfigs(
+      cfg_.use_plugin_optimizers() != RewriterConfig::OFF, device_types);
+  if (!cfg_.disable_model_pruning() && !plugin_configs.disable_model_pruning) {
     optimizers->push_back(MakeUnique<ModelPruner>());
   }
-  if (cfg_.implementation_selector() != RewriterConfig::OFF) {
+  if (cfg_.implementation_selector() != RewriterConfig::OFF &&
```

Comment:
Maybe move this repeated pattern to macro?

```
cfg_.implementation_selector() != RewriterConfig::OFF &&
  plugin_configs.toggle_config["implementation_selector"]
```

Reply:
The pattern is not exactly the same for different optimizers, so I use 6 macros here.
~~~cpp
#define USER_IS_ON(CFG) cfg_.CFG() == RewriterConfig::ON
#define USER_NOT_OFF(CFG) cfg_.CFG() != RewriterConfig::OFF
#define PLUGIN_IS_ON(CFG) plugin_configs.toggle_config[#CFG] == RewriterConfig::ON
#define PLUGIN_NOT_OFF(CFG) plugin_configs.toggle_config[#CFG] != RewriterConfig::OFF
#define BOTH_IS_ON(CFG) USER_IS_ON(CFG) && PLUGIN_IS_ON(CFG)
#define BOTH_NOT_OFF(CFG) USER_NOT_OFF(CFG) && PLUGIN_NOT_OFF(CFG)
~~~